feat: langchain llm message attributes

axiomofjoy opened the pull request:
- [ ] implements `llm.messages` semantic convention for langchain tracer
- [ ] adds unit tests for completions and chat completions examples

RogerHYang commented on src/phoenix/trace/langchain/tracer.py:
```diff
@@ -170,7 +195,50 @@ def _retrieval_documents(
+    For reference, see https://github.com/langchain-ai/langchain/pull/4499.
+    """
+
+    def on_chat_model_start(
```
why not add the method directly to OpenInferenceTracer? wouldn't it be simpler that way?

axiomofjoy replied:
Addressed in [648aa51](https://github.com/Arize-ai/phoenix/pull/1410/commits/648aa51e5379854eae7ff51befd790ebe2defaef)

axiomofjoy commented:
# Learnings

The OpenAI chat completions API takes in a sequence of messages. By default, it output a single message in response.

The OpenAI completions API takes in either a single prompt or a batch of prompts. By default, it outputs one completion for each prompt it receives.

Both the completions and chat completions API support a parameter `n` that defaults to 1 that determines the number the number of "choices" to return to the user. Each input (message sequence or prompt) receives `n` outputs. If `n` equals 3, the chat completions endpoint will return three different messages, each message being a possible response to the input sequence of messages. If `n` equals 3 and the completions endpoint receives a batch of two input prompts, it will respond with six completions, three possible completions for each prompt.

LlamaIndex ignores this complexity. The model interface accepts a single message sequence in the case of chat models or a single prompt in the case of completion models. The `n` parameter is not supported, so only one output is computed. One input, one output.

LangChain supports the full complexity of the API. The completions model interface accepts a batch of inputs prompts and also accepts the parameter `n`. When a user passes multiple prompts as input, LangChain's callback system currently treats each prompt as belonging to its own span, even though all prompts were completed in a single request. For example, if we complete two prompts with `n = 3`, we wind up with two spans, each span containing a single prompt as input and three choices of completions for output. We have two spans, but only one request.

It makes sense to have two spans in this situation if what you care about evaluating is the quality of individual prompt-response pairs (e.g., evaluating the output for toxicity). It makes less sense if what you care about is request latency.

Here are a few code snippets as illustration.

## Chat Completions

Two message sequences, two requests, `n = 3`, three choices

```python
import logging
import sys
from langchain.chat_models import ChatOpenAI
from langchain.schema.messages import ChatMessage

logging.basicConfig(level=logging.DEBUG, stream=sys.stdout)

llm = ChatOpenAI(model_name="gpt-3.5-turbo", n=3)
batches = [
    [
        ChatMessage(role="system", content="Who won the World Cup in 2018?"),
    ],
    [
        ChatMessage(
            role="system",
            content="Who is the most decorated Olympic swimmer of all time?",
        ),
    ],
]
result = llm.generate(batches)

for message_sequence_index, (messages, generations) in enumerate(
    zip(batches, result.generations)
):
    for generation_index, generation in enumerate(generations):
        print(
            f"Message sequence {message_sequence_index} Generation {generation_index}"
        )
        print("==========================")
        print()
        print(messages)
        print()
        print(generation.text)
        print()
```

### Output:

```
DEBUG:openai:message='Request to OpenAI API' method=post path=https://api.openai.com/v1/chat/completions
DEBUG:openai:api_version=None data='{"messages": [{"role": "system", "content": "Who won the World Cup in 2018?"}], "model": "gpt-3.5-turbo", "max_tokens": null, "stream": false, "n": 3, "temperature": 0.7}' message='Post details'
DEBUG:urllib3.util.retry:Converted retries value: 2 -> Retry(total=2, connect=None, read=None, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.openai.com:443
DEBUG:urllib3.connectionpool:https://api.openai.com:443 "POST /v1/chat/completions HTTP/1.1" 200 None
DEBUG:openai:message='OpenAI API response' path=https://api.openai.com/v1/chat/completions processing_ms=638 request_id=a3e07084fe534465a15ab5edaee75915 response_code=200
DEBUG:openai:message='Request to OpenAI API' method=post path=https://api.openai.com/v1/chat/completions
DEBUG:openai:api_version=None data='{"messages": [{"role": "system", "content": "Who is the most decorated Olympic swimmer of all time?"}], "model": "gpt-3.5-turbo", "max_tokens": null, "stream": false, "n": 3, "temperature": 0.7}' message='Post details'
DEBUG:urllib3.connectionpool:https://api.openai.com:443 "POST /v1/chat/completions HTTP/1.1" 200 None
DEBUG:openai:message='OpenAI API response' path=https://api.openai.com/v1/chat/completions processing_ms=1931 request_id=bf8d0a00b7e0b66772b2bff0a541c8bb response_code=200
Message sequence 0 Generation 0
==========================

[ChatMessage(content='Who won the World Cup in 2018?', additional_kwargs={}, role='system')]

France won the FIFA World Cup in 2018.

Message sequence 0 Generation 1
==========================

[ChatMessage(content='Who won the World Cup in 2018?', additional_kwargs={}, role='system')]

The French national team won the FIFA World Cup in 2018.

Message sequence 0 Generation 2
==========================

[ChatMessage(content='Who won the World Cup in 2018?', additional_kwargs={}, role='system')]

France won the World Cup in 2018.

Message sequence 1 Generation 0
==========================

[ChatMessage(content='Who is the most decorated Olympic swimmer of all time?', additional_kwargs={}, role='system')]

The most decorated Olympic swimmer of all time is Michael Phelps from the United States. He has won a total of 28 Olympic medals, including 23 gold medals.

Message sequence 1 Generation 1
==========================

[ChatMessage(content='Who is the most decorated Olympic swimmer of all time?', additional_kwargs={}, role='system')]

Michael Phelps is the most decorated Olympic swimmer of all time. He has won a total of 23 gold medals, 3 silver medals, and 2 bronze medals in his Olympic career.

Message sequence 1 Generation 2
==========================

[ChatMessage(content='Who is the most decorated Olympic swimmer of all time?', additional_kwargs={}, role='system')]

The most decorated Olympic swimmer of all time is Michael Phelps from the United States. He has won a total of 28 Olympic medals, including 23 gold medals, between 2004 and 2016.
```

## Completions API

Two prompts, one request, `n = 3`, six completions

### Code:

```python
import logging
import sys
from langchain.llms import OpenAI

logging.basicConfig(level=logging.DEBUG, stream=sys.stdout)

llm = OpenAI(model_name="gpt-3.5-turbo-instruct", temperature=0.75, n=3)
prompts = [
    "Who won the World Cup in 2018?",
    "Who is the most decorated Olympic swimmer of all time?",
]
result = llm.generate(prompts)

for prompt_index, (prompt, generations) in enumerate(zip(prompts, result.generations)):
    for generation_index, generation in enumerate(generations):
        print(f"Prompt {prompt_index} Generation {generation_index}")
        print("==========================")
        print()
        print(prompt)
        print()
        print(generation.text)
        print()
```

### Output:

```
DEBUG:openai:message='Request to OpenAI API' method=post path=https://api.openai.com/v1/completions
DEBUG:openai:api_version=None data='{"prompt": ["Who won the World Cup in 2018?", "Who is the most decorated Olympic swimmer of all time?"], "model": "gpt-3.5-turbo-instruct", "temperature": 0.75, "max_tokens": 256, "top_p": 1, "frequency_penalty": 0, "presence_penalty": 0, "n": 3, "logit_bias": {}}' message='Post details'
DEBUG:urllib3.util.retry:Converted retries value: 2 -> Retry(total=2, connect=None, read=None, redirect=None, status=None)
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.openai.com:443
DEBUG:urllib3.connectionpool:https://api.openai.com:443 "POST /v1/completions HTTP/1.1" 200 None
DEBUG:openai:message='OpenAI API response' path=https://api.openai.com/v1/completions processing_ms=784 request_id=ebc40b8b7b74b1539dd60c3953e11cba response_code=200
Prompt 0 Generation 0
==========================

Who won the World Cup in 2018?



France won the World Cup in 2018.

Prompt 0 Generation 1
==========================

Who won the World Cup in 2018?



France won the World Cup in 2018.

Prompt 0 Generation 2
==========================

Who won the World Cup in 2018?



France won the 2018 FIFA World Cup, defeating Croatia 4-2 in the final.

Prompt 1 Generation 0
==========================

Who is the most decorated Olympic swimmer of all time?



The most decorated Olympic swimmer of all time is Michael Phelps, with a total of 28 Olympic medals (23 gold, 3 silver, and 2 bronze). 

Prompt 1 Generation 1
==========================

Who is the most decorated Olympic swimmer of all time?



The most decorated Olympic swimmer of all time is Michael Phelps of the United States, with a total of 28 medals (23 gold, 3 silver, and 2 bronze).

Prompt 1 Generation 2
==========================

Who is the most decorated Olympic swimmer of all time?



The most decorated Olympic swimmer of all time is Michael Phelps. He has won a total of 28 Olympic medals, including 23 gold medals. He also holds the record for the most gold medals won at a single Olympics (8 gold medals at the 2008 Beijing Olympics).
```

We recently switched from a singular `llm.prompt` to a plural `llm.prompts` semantic attribute under the assumption that LangChain would have multiple prompts in a single LLM span when using the completions API. From what I have seen so far, it seems like that's not the convention LangChain is following.

axiomofjoy commented on src/phoenix/trace/langchain/tracer.py:
```diff
@@ -234,3 +281,43 @@ def _persist_run(self, run: Run) -> None:
         except Exception:
             logger.exception("Failed to convert run to spans")
+
+    def on_chat_model_start(
```
The way this currently works is that the callback system first tries calling `on_chat_model_start` for chat models. If that hook is not implemented, it falls back to `on_llm_start`. Somewhat confusingly, there is no `on_chat_model_end` hook, only a `on_llm_end` hook that is call regardless of whether `on_chat_model_start` is implemented.

https://github.com/langchain-ai/langchain/blob/f0198354d93e7ba8b615b8fd845223c88ea4ed2b/libs/langchain/langchain/callbacks/manager.py#L381
